src/eduid_userdb/actions/action.py

```diff
@@ -0,0 +1,204 @@
+    """
+    Generic eduID action object.
+
+    If the `data' argument is None (default), the Action will be created from the other
+    keyword arguments.
+
+    :param data: MongoDB document representing an action
+    :param action_id: Unique identifier for the action
+    :param user_oid: User id
+    :param action_type: What action to perform
+    :param preference: Used to sort actions
+    :param session: IdP session identifier
+    :param params: Parameters for action
+    :param raise_on_unknown: Raise exception on unknown data or not
+
+    :type data: dict | None
+    :type action_id: bson.ObjectId | str | None
+    :type action_type: str | None
+    :type preference: int | None
+    :type session: str | None
+    :type params: dict | None
+    :type raise_on_unknown: bool
+    """
+    def __init__(self, action_id = None, user_oid = None, action_type = None, preference = None, session = None,
+                 params = None, data = None, raise_on_unknown = True):
+        self._data_in = copy.deepcopy(data)  # to not modify callers data
+        self._data = dict()
+
+        if self._data_in is None:
+            self._data_in = dict(_id = action_id,
+                                 user_oid = user_oid,
+                                 action = action_type,
+                                 preference = preference,
+                                 session = session or '',
+                                 params = params or {})
+
+        self._parse_check_invalid_actions()
+
+        # ensure _id is always an ObjectId
+        _id = self._data_in.pop('_id', None)
+        if _id is None:
+            _id = bson.ObjectId()
+        elif not isinstance(_id, bson.ObjectId):
+            _id = bson.ObjectId(_id)
+
+        # things without setters
+        self._data['_id'] = _id
+        self._data['user_oid'] = self._data_in.pop('user_oid')
+        self._data['action'] = self._data_in.pop('action')
+        self._data['preference'] = self._data_in.pop('preference', 100)
+        self._data['session'] = self._data_in.pop('session', '')
+        self._data['params'] = self._data_in.pop('params', {})
+
+        if len(self._data_in) > 0:
+            if raise_on_unknown:
+                raise ActionHasUnknownData('Action {!s} unknown data: {!r}'.format(
+                    self.action_id, self._data_in.keys()
+                ))
+            # Just keep everything that is left as-is
+            self._data.update(self._data_in)
+
+        del self._data_in
+
+    def _parse_check_invalid_actions(self):
+        """"
+        Part of __init__().
+
+        Check actions that can't be loaded for some known reason.
+        """
+        if self._data_in.get('user_oid') is None:
+            raise ActionMissingData('Action {!s} has no user_oid'.format(
+                self._data_in.get('_id')))
+        if self._data_in.get('action') is None:
+            raise ActionMissingData('Action {!s} has no action'.format(
```

Comment:
Yes, Action has no action sounds weird, I have committed the change suggested by Johan, directly in master.